sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -699,6 +699,24 @@ object CollapseProject extends Rule[LogicalPlan] {
         agg.copy(aggregateExpressions = buildCleanedProjectList(
           p.projectList, agg.aggregateExpressions))
       }
+    case p1 @ Project(_, g @ GlobalLimit(_, l @ LocalLimit(_, p2: Project))) =>
+      if (haveCommonNonDeterministicOutput(p1.projectList, p2.projectList)) {
+        p1
+      } else {
+        val newProjectList = buildCleanedProjectList(p1.projectList, p2.projectList)
+        g.copy(child = l.copy(child = p2.copy(projectList = newProjectList)))
+      }
+    case p1 @ Project(_, l @ LocalLimit(_, p2: Project)) =>
+      if (haveCommonNonDeterministicOutput(p1.projectList, p2.projectList)) {
+        p1
+      } else {
+        val newProjectList = buildCleanedProjectList(p1.projectList, p2.projectList)
+        l.copy(child = p2.copy(projectList = newProjectList))
+      }
+    case Project(list, r @ Repartition(_, _, p: Project)) if list.forall(_.deterministic) =>
+      r.copy(child = p.copy(projectList = buildCleanedProjectList(list, p.projectList)))
```

Comment:
Is it possible that this can increase the shuffled data?

For example, we apply a deterministic UDF taking two input in the project which produces like 10 columns as output after repartition. With this rule, this UDF will be moved before the repartition which can result more shuffle, right?

Thanks!

Reply:
Oops. My bad. I used too broad pattern. I'll narrow down specifically for the renaming cases.

Reply:
Could you leave the comment about why the deterministic handling is different between in `Repartition/Sample` and the others?

Reply:
We don't need to narrow down the two Limit cases above, too? In the same case @dbtsai suggested, this rule possibly increases processing data in limit queries? `Project(udfs generates many data) <- limit <- project`.

Reply:
Could you leave a comment on the last commit? This one is old one.